src/generic/gridsel.cpp

```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray cells;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; n++ )
+    {
+        const wxGridBlockCoords& block = m_selection[n];
+        if ( block.GetTopRow() == block.GetBottomRow() &&
+             block.GetLeftCol() == block.GetRightCol() )
+        {
+            cells.Add(block.GetTopLeft());
+        }
+    }
+    return cells;
+}
+
+wxGridCellCoordsArray wxGridSelection::GetBlockSelectionTopLeft() const
+{
+    // return blocks only in wxGridSelectCells selection mode
+    if ( m_selectionMode != wxGrid::wxGridSelectCells )
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray coords;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; n++ )
+    {
+        coords.Add(m_selection[n].GetTopLeft());
+    }
+    return coords;
+}
+
+wxGridCellCoordsArray wxGridSelection::GetBlockSelectionBottomRight() const
+{
+    if ( m_selectionMode != wxGrid::wxGridSelectCells )
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray coords;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; n++ )
+    {
+        coords.Add(m_selection[n].GetBottomRight());
+    }
+    return coords;
+}
+
+wxArrayInt wxGridSelection::GetRowSelection() const
+{
+    if ( m_selectionMode == wxGrid::wxGridSelectColumns )
+        return wxArrayInt();
+
+    wxIntSet uniqueRows;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; ++n )
+    {
+        const wxGridBlockCoords& block = m_selection[n];
+        if ( block.GetLeftCol() == 0 &&
+             block.GetRightCol() == m_grid->GetNumberCols() - 1 )
+        {
+            for ( int r = block.GetTopRow(); r <= block.GetBottomRow(); ++r )
+            {
+                uniqueRows.insert(r);
+            }
+        }
+    }
+
+    wxArrayInt result;
+    result.Alloc(uniqueRows.size());
+    wxIntSet::iterator it;
+    for( it = uniqueRows.begin(); it != uniqueRows.end(); ++it )
+    {
+        result.Add(*it);
+    }
+    return result;
+}
+
+wxArrayInt wxGridSelection::GetColSelection() const
+{
+    if ( m_selectionMode == wxGrid::wxGridSelectRows )
+        return wxArrayInt();
+
+    wxIntSet uniqueCols;
+    const size_t count = m_selection.size();
+    for ( size_t n = 0; n < count; ++n )
+    {
+        const wxGridBlockCoords& block = m_selection[n];
+        if ( block.GetTopRow() == 0 &&
+             block.GetBottomRow() == m_grid->GetNumberRows() - 1 )
+        {
+            for ( int c = block.GetLeftCol(); c <= block.GetRightCol(); ++c )
+            {
+                uniqueCols.insert(c);
             }
         }
     }
 
-    // No need to touch selected rows, unless we removed _all_
-    // columns, in this case, we remove all rows from the selection.
-    if ( !m_grid->GetNumberCols() )
-        m_rowSelection.Clear();
+    wxArrayInt result;
+    result.Alloc(uniqueCols.size());
```

Comment:
```suggestion
    result.reserve(uniqueCols.size());
```
Prefer using standard-like functions when possible.